python/pyarrow/_dataset.pyx

```diff
@@ -96,27 +96,33 @@ def _get_parquet_symbol(name):
     return _dataset_pq and getattr(_dataset_pq, name)
 
 
-cdef CFileSource _make_file_source(object file, FileSystem filesystem=None):
+cdef CFileSource _make_file_source(object file, FileSystem filesystem=None, int64_t file_size=-1):
```

Comment:
> Instead of coercing file_size here and also in make_fragment, could we instead allow this function to accept None and move it all here?

Hmm this doesn't compile for me

Reply:
> Instead of coercing file_size here and also in make_fragment, could we instead allow this function to accept None and move it all here?

I tried that initially but couldn't get it to work, hence the awkward integer argument. Sadly the earlier conversation is lost because I have force-pushed since that :(

Reply:
> Also, based on call sites below it looks like you mean to require kwargs for filesystem and file_size

Hmm this doesn't compile for me (* is not allowed)

Reply:
Huh, I didn't know that was disallowed for cdef functions... maybe it makes parsing pointer args harder?

Re the default arg, I made this metaPR https://github.com/eeroel/arrow/pull/1 which compiled and passed your test case on my machine

Reply:
Thanks a lot, much cleaner now! Can't for the life of me remember why I couldn't get it to work earlier 🤔